### ISSUE TITLE
TCC.xsl:48-50: `directly-related-pairs` currently don't...

### ISSUE THREAD

**Issue by 0pdd** [yegor256/jpeek#120]
The puzzle `9-57e7e31a` from #9 has to be resolved:

https://github.com/yegor256/jpeek/blob/b625770426cbfa670c987c135c0b3c98b7a24c8d/src/main/resources/org/jpeek/metrics/TCC.xsl#L48-L50

The puzzle was created by Kapralov Sergey on 26-Jan-18. 

Estimate: 30 minutes, role: IMP.

If you have any technical questions, don't ask me, submit new tickets instead. The task will be "done" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

**Comment by 0crat** [comment]
@yegor256[/z](https://www.0crat.com/u/yegor256) please, pay attention to this issue

**Comment by 0crat** [comment]
Job [#120](https://github.com/yegor256/jpeek/issues/120) is now in scope, role is `DEV`

**Comment by 0crat** [comment]
The job [#120](https://github.com/yegor256/jpeek/issues/120) assigned to @llorllale[/z](https://www.0crat.com/u/llorllale), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/450314418). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by llorllale** [comment]
My own notes:

* They keep using the term "instance variable" but as far as I can see they do not actually differentiate it from class variables
* They deliberately exclude ctors from the analysis because "It will generally access all instance variables in the class, and thus, share instance variables with virtually all other methods. Constructors create connections between methods even if the methods do not have any other relationships."
* They deliberately exclude "non-visible" methods from outside the class because "Class cohesion refers the relatedness of visible components of the class which represent its functionality. Class cohesion is the degree that those components are related. In our model of class cohesion, invisible components of a class are included only indirectly through the visible components."
* They leave open the possibility for including inherited methods and instance variables. We are currently including the FQN of methods used (will do the same for fields in the near future), but we also need to know if the owning class is a parent of the one under analysis. I opened #187 for this.
* Formula is TCC(C) = NDC(C)/NP(C), where NP is the max number of possible connections (N*(N-1)/2), and NDC is the number of directly connected methods
* The direct connectivity between methods is determined from the class abstraction. If there exists one or more common instance variables between two method abstractions then the two corresponding methods are directly connected.
* They note the following when they applied their cohesion measure: "Overloaded methods within the same class are treated as one method."
* Class cohesion measures for visible methods defined only within the class are also useful, because the measures are not affected by the cohesion of a superclass. This should be configurable?
* Quote: "Our results show that the classes that are heavily reused via inheritance exhibit lower cohesion".

**Comment by llorllale** [comment]
Based on my notes above I've decided the following:

* TCC does not care about relations between methods via method calls, so the request to do so in the puzzle will be discarded
* Considering the consistent use of the term "instance variable" and their consideration of inheritance, etc., I am going to only consider class attributes that are not static.
* I am going to exclude ctors
* I am going to exclude private methods
* I am going to leave a puzzle to come back and fix TCC after #187 is closed. Inclusion of inherited attrs and methods should be configurable.

**Comment by 0pdd** [comment]
The puzzle `9-57e7e31a` has disappeared from the source code, that's why I closed this issue.

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job, as in [§30](http://datum.zerocracy.com/pages/policy.html#30)

**Comment by llorllale** [comment]
@yegor256 missed points+payment on this one

**Comment by llorllale** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/120#issuecomment-366520819))

@llorllale This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#120](https://github.com/yegor256/jpeek/issues/120) is [in scope](http://datum.zerocracy.com/pages/policy.html#14) for 21days
  * The role is `DEV`
  * The job is assigned to @llorllale[/z](https://www.0crat.com/u/llorllale) for [4days](http://datum.zerocracy.com/pages/policy.html#8)
  * There is a monetary reward attached
  * @llorllale[/z](https://www.0crat.com/u/llorllale) will get 50% of their hourly rate for this job
  * The job doesn't have any [impediments](http://datum.zerocracy.com/pages/policy.html#9)
  * The [budget](http://datum.zerocracy.com/pages/policy.html#4) is 30 minutes/points
  * These users are banned and won't be assigned:
    * @0pdd[/z](https://www.0crat.com/u/0pdd): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C7JGJ00DP?q=%7Bjob%3A%27gh%3Ayegor256%2Fjpeek%23120%27%7D) (restricted area)

**Comment by 0crat** [comment]
@llorllale[/z](https://www.0crat.com/u/llorllale) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by llorllale** [comment]
@ypshenychka seems like you must also review the associated PR? (#182). See this points in the policy:

> Pull request title explains the problem it is fixing.
Pull request description explains the solution proposed and contains a link to the original ticket it is related to.
The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.
All problems found by the code reviewer were addressed by the pull request author.

**Comment by 0crat** [comment]
QA review completed: +15 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka), total is [+180](https://www.0crat.com/u/ypshenychka)

**Comment by llorllale** [comment]
@ypshenychka nevermind

**Comment by 0crat** [comment]
Order was finished, quality was "good": +35 points just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale), total is [+655](https://www.0crat.com/u/llorllale)

**Comment by 0pdd** [comment]
@0pdd these puzzles are still not solved: [193](https://github.com/yegor256/jpeek/issues/193), [194](https://github.com/yegor256/jpeek/issues/194), [195](https://github.com/yegor256/jpeek/issues/195)

**Comment by 0pdd** [comment]
@0pdd these puzzles are still not solved: [193](https://github.com/yegor256/jpeek/issues/193), [194](https://github.com/yegor256/jpeek/issues/194), [195](https://github.com/yegor256/jpeek/issues/195)

**Comment by 0pdd** [comment]
@0pdd these puzzles are still not solved: [193](https://github.com/yegor256/jpeek/issues/193), [194](https://github.com/yegor256/jpeek/issues/194), [195](https://github.com/yegor256/jpeek/issues/195)

**Comment by 0pdd** [comment]
@0pdd 2 puzzles [#193](https://github.com/yegor256/jpeek/issues/193), [#194](https://github.com/yegor256/jpeek/issues/194) are still not solved; solved: [#195](https://github.com/yegor256/jpeek/issues/195).

**Comment by 0pdd** [comment]
@0pdd 2 puzzles [#495](https://github.com/cqfn/jpeek/issues/495), [#496](https://github.com/cqfn/jpeek/issues/496) are still not solved.

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/cqfn/jpeek/issues/120#issuecomment-366520819))

@llorllale This is what I know about this job in C7JGJ00DP, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#120](https://github.com/cqfn/jpeek/issues/120) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C7JGJ00DP?q=%7Bjob%3A%27gh%3Acqfn%2Fjpeek%23120%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/cqfn/jpeek/issues/120#issuecomment-366802067))

@ypshenychka You @ypshenychka[/z](https://www.0crat.com/u/ypshenychka) need to have one of these roles in order to do "Complete `QA` review": `ARC`, `QA`; I'm sorry to say this, but at the moment you've got no roles in this project

**Comment by 0pdd** [comment]
@0pdd 2 puzzles [#581](https://github.com/cqfn/jpeek/issues/581), [#582](https://github.com/cqfn/jpeek/issues/582) are still not solved; solved: [#495](https://github.com/cqfn/jpeek/issues/495), [#496](https://github.com/cqfn/jpeek/issues/496).
